src/pymor/reductors/mt.py

```diff
@@ -0,0 +1,169 @@
+
+            - `'eig'`: use scipy.linalg.eig algorithm
+            - `'samdp'`: find dominant poles using samdp algorithm
+        projection
+            Projection method:
+
+            - `'orth'`: projection matrices are orthogonalized with
+              respect to the Euclidean inner product
+            - `'biorth'`: projection matrices are biorthogolized with
+              respect to the E product
+        symmetric
+            True if Operator A is symmetric and E is symmetric, positive
+            definite, False if not.
+        method_options
+            Optional dict with more options for the samdp algorithm.
+        which
+            A string specifying which `k` eigenvalues and eigenvectors to
+            compute when using the eig decomposition, default is "LR":
+
+            - `'SM'`: select eigenvalues with smallest magnitude
+            - `'LR'`: select eigenvalues with largest real part
+            - `'NR'`: select eigenvalues with largest norm(residual) / abs(Re(pole))
+            - `'NS'`: select eigenvalues with largest norm(residual) / abs(pole)
+            - `'NM'`: select eigenvalues with largest norm(residual)
+             
+        Returns
+        -------
+        rom
+            Reduced-order model.
+        """
+        assert 0 < r < self.fom.order
+        assert projection in ('orth', 'biorth')
+        assert decomposition in ('eig', 'samdp')
+        assert method_options is None or isinstance(method_options, dict)
+        if not method_options:
+            method_options = {}
+
+        self.V = self.fom.B.as_range_array().empty(reserve=r)
+        self.W = self.fom.C.as_source_array().empty(reserve=r)
+
+        if decomposition == 'eig':
+            if self.fom.A.sparse:
+                self.logger.warning('Converting operator A to a NumPy array.')
+            A = to_matrix(self.fom.A, format='dense')
+            if not isinstance(self.fom.E, IdentityOperator):
+                if self.fom.E.sparse:
+                    self.logger.warning('Converting operator E to a NumPy array.')
+            E = None if isinstance(self.fom.E, IdentityOperator) else to_matrix(self.fom.E, format='dense')
+
+            if symmetric:
+                poles, ev_r = spla.eig(A, E, right=True)
+                rev = self.fom.A.source.from_numpy(ev_r.T)                
+                lev = rev.copy()
+            else:
+                poles, ev_l, ev_r = spla.eig(A, E, left=True, right=True)
+                rev = self.fom.A.source.from_numpy(ev_r.T)
+                lev = self.fom.A.source.from_numpy(ev_l.T)
+            if which == 'SM':
+                idx = np.argsort(np.abs(poles))
+            elif which == 'LR':
+                idx = np.argsort(-poles.real)
+            else:
+                absres = np.empty(len(poles))
+                for i in range(len(poles)):
+                    b_norm = spla.norm(lev[i].inner(self.fom.B.as_range_array()), ord=2)
```

Comment:
It looks mathematically correct, but it should be possible to avoid using `as_range_array`. Try using `self.fom.B.apply_adjoint` and the `norm` method of `VectorArrays`.

Similar for `c_norm` below.